### ISSUE TITLE
Strange behaviors if curry function with named arguments

### ISSUE THREAD

**Issue by zydmayday** [ramda/ramda#3284]
I am wondering how ramda support functions with named arguments.

Here are 2 examples.

[1 normal argument, 1 named argument](https://ramdajs.com/repl/#?const%20f%20%3D%20%28a%2C%20b%3D2%29%20%3D%3E%20%5Ba%2C%20b%5D%0Aconst%20g%20%3D%20curry%28f%29%0A%0Ag%28__%2C%20200%29%281%29%20%2F%2F%20200%20will%20be%20ignored)
```js
const f = (a, b=2) => [a, b]
const g = curry(f)

g(__, 200)(1) // [1, 2], 200 will be ignored
```

[2 normal arguments, 1 named argument](https://ramdajs.com/repl/#?const%20f%20%3D%20%28a%2C%20b%2C%20c%3D3%29%20%3D%3E%20%5Ba%2C%20b%2C%20c%5D%0Aconst%20g%20%3D%20curry%28f%29%0A%0Ag%28__%2C%20__%2C%20300%29%281%29%20%2F%2F%20%5B1%2C%20__%2C%20300%5D%0Ag%28__%2C%20__%2C%20300%29%281%2C%202%29%20%2F%2F%20%5B1%2C%202%2C%20300%5D%0Ag%28__%2C%20__%2C%20300%29%28__%2C%202%29%20%2F%2F%20%5B__%2C%202%2C%20300%5D%0Ag%28__%2C%20__%2C%20300%29%28__%2C%20__%29%281%29%20%2F%2F%20%5B1%2C%20__%2C%20300%5D%0Ag%28__%2C%20__%2C%20300%29%28__%2C%20__%29%281%2C%202%29%20%2F%2F%20%5B1%2C%202%2C%20300%5D)
```js
const f = (a, b, c=3) => [a, b, c]
const g = curry(f)

// Not like above example, 300 is not ignored here.
// R.__ is also not ignored sometimes.
g(__, __, 300)(1) // [1, __, 300]
g(__, __, 300)(1, 2) // [1, 2, 300]
g(__, __, 300)(__, 2) // [__, 2, 300]
g(__, __, 300)(__, __)(1) // [1, __, 300]
g(__, __, 300)(__, __)(1, 2) // [1, 2, 300]
```

So maybe ramda not fully support named arguments for now?
Thanks!

**Comment by CrossEye** [comment]
The problem here is that Ramda cannot work particularly well with optional arguments.  When you use a default argument, it becomes optional and the arity is reduced.  Whereas

```js
((a, b) => a + b) .length //=> 2
```

if we default an argument, we see:

```js
((a, b = 1) => a + b) .length //=> 1
```

Ramda uses the function's arity (exposed as `length`) to determine how the currying and placeholder usage works.

We do try to make it clear that Ramda doesn't work well with optional values.  If you can think of a way to make that more prominent, please let us know.

If you  work through the implementations of https://github.com/ramda/ramda/blob/v0.28.0/source/internal/_curryN.js and https://github.com/ramda/ramda/blob/v0.28.0/source/curryN.js, you can see the details on how we get the specific values you see.  But the main point is simply that Ramda does not work well with optional values.

**Comment by customcommander** [comment]
The documentation for `curry` has been updated for that reason in #3244 however it needs to be deployed. (Although I really need to address this [comment](https://github.com/ramda/ramda/commit/aaffe722cadcb366aadf1f547f99195cbdef31fe#comments) :P)

I've got some free time end of week so I should be able to try to isolate the breaking changes in a `v1.0` branch so we could release a non-breaking `v0.28.2` with this and a couple of other things people are waiting for. No promise but I'll try!
